_posts/2017-11-30-totw-130.md

```diff
@@ -0,0 +1,185 @@
+interesting less-followed option that an entire top-level package
+(search/...) could choose a single flat top-level namespace. Following on
+the heels of Java's package naming requirements, this makes a lot of sense: we
+want to be able to uniquely identify symbols in C++ and we want there to be
+uniqueness and consistency in namespace choice.
+
+Except in actuality, we don't. We just didn't realize for almost a decade.
+
+## Name lookup
+
+Let's start with how name lookup works in C++ and how it's different from Java.
+
+``` c++
+namespace foo {
+namespace bar {
+void f() {
+  Baz b;
+}
+}
+}
+```
+
+In C++, lookup on an unqualified name (`Baz`) will search expanding scopes for
+a symbol of the same name: first in `f` (the function), then in `bar`, then in
+`foo`, then in the global namespace.
+
+
+In Java, there is no such thing as an unqualified symbol: either a symbol is
+a qualified name
+
+``` java
+public void f() {
+  com.google.foo.bar.Baz b = new com.google.foo.bar.Baz();
+}
+```
+
+or it is imported, either as a single package member or via wildcard:
+
+``` java
+import com.google.foo.bar.Baz;
+import com.google.foo.bar.*;
+```
+
+In no case is `Baz` looked for outside of the package that is explicitly
+provided: wildcards don't descend into child packages, nor is search extended
+into parent packages. As it turns out, the difference in how parent
+packages/namespaces is handled is fundamental to why structural namespace naming
+(making the namespace structure match the package hierarchy) is a mistake.
+
+## The problem
+
+The fundamental problem for building namespaces out of packages is that we
+rarely rely on fully-qualified lookup in C++, normally writing `std::unique_ptr`
+rather than `::std::unique_ptr`. Coupled with lookup in enclosing namespaces,
+this means that for code in a deeply nested package
+(`::division::section::team::subteam::project`, for example) any symbol that is not
+fully qualified (`std::unique_ptr`) can in fact reference any of
+
+*   `::std::unique_ptr`
+*   `::division::std::unique_ptr`
+*   `::division::section::std::unique_ptr`
+*   `::division::section::team::std::unique_ptr`
+*   `::division::section::team::subteam::std::unique_ptr`
+*   `::division::section::team::subteam::project::std::unique_ptr`
+
+And what's worse: unqualified search starts at the bottom of that list *and
+stops as soon as there is a namespace match*. This means that your build can be
+broken if any of your transitive includes add a previously unused namespace that
+matches the leading namespace of any symbol you use out of an unqualified
+namespace. Strictly speaking, this doesn't even have to be a build break: if
+someone adds something with a matching name and a syntactically-compatible API,
+the implementation of that API may be completely incompatible and cause
+widespread havoc at runtime. Obviously this isn't too bad with `std` - nobody
+should ever be adding a nested namespace `std`, but what about more common
```

Comment:
\`std\`, -> \`std\`  -

(You introduce the aside with a dash, so you need to close it with one)